File: internal/middleware/ratelimit.go
package middleware

import (
	"net"
	"net/http"
	"strconv"
	"strings"
	"sync"
	"time"

	"github.com/rpattn/raalisence/internal/config"
)

type bucket struct {
	tokens     float64
	lastRefill time.Time
}

type limiter struct {
	mu        sync.Mutex
	buckets   map[string]*bucket
	rps       float64       // tokens per second
	burst     float64       // max tokens
	ttl       time.Duration // idle bucket eviction
	lastSweep time.Time
}

func newLimiter(rps float64, burst int, ttl time.Duration) *limiter {
	return &limiter{
		buckets:   make(map[string]*bucket),
		rps:       rps,
		burst:     float64(burst),
		ttl:       ttl,
		lastSweep: time.Now(),
	}
}

func (l *limiter) allow(key string) (ok bool, remaining int, retryAfter time.Duration) {
	l.mu.Lock()
	defer l.mu.Unlock()

	now := time.Now()
	// periodic sweep of stale buckets
	if now.Sub(l.lastSweep) > l.ttl {
		for k, b := range l.buckets {
			if now.Sub(b.lastRefill) > l.ttl {
				delete(l.buckets, k)
			}
		}
		l.lastSweep = now
	}

	b := l.buckets[key]
	if b == nil {
		b = &bucket{tokens: l.burst, lastRefill: now}
		l.buckets[key] = b
	}

	// refill
	elapsed := now.Sub(b.lastRefill).Seconds()
	b.tokens = mathMin(l.burst, b.tokens+elapsed*l.rps)
	b.lastRefill = now

	if b.tokens >= 1.0 {
		b.tokens -= 1.0
		return true, int(b.tokens), 0
	}

	missing := 1.0 - b.tokens
	retryAfter = time.Duration(missing / l.rps * float64(time.Second))
	return false, int(b.tokens), retryAfter
}

func mathMin(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

// WithRateLimit applies a simple token bucket rate limit per client.
// Keying strategy:
// - Admin endpoints (/issue, /revoke) are keyed by admin token (so two admins behind the same IP aren't unfairly throttled).
// - Other endpoints keyed by client IP (first X-Forwarded-For hop if present, else RemoteAddr).
func WithRateLimit(cfg *config.Config, next http.Handler) http.Handler {
	// Defaults (tweak as you like or expose in config)
	fast := newLimiter(5, 10, 10*time.Minute) // validate/heartbeat
	admin := newLimiter(1, 3, 10*time.Minute) // issue/revoke
	deflt := newLimiter(2, 5, 10*time.Minute) // everything else

	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
		var l *limiter
		key := rateKey(cfg, r)
		switch r.URL.Path {
		case "/api/v1/licenses/validate", "/api/v1/licenses/heartbeat":
			l = fast
		case "/api/v1/licenses/issue", "/api/v1/licenses/revoke":
			l = admin
		default:
			l = deflt
		}

		ok, remaining, retry := l.allow(key)
		w.Header().Set("RateLimit-Limit", "1")
		w.Header().Set("RateLimit-Remaining", strconv.Itoa(remaining))
		if !ok {
			if retry < 0 {
				retry = 0
			}
			w.Header().Set("Retry-After", strconv.FormatInt(int64(retry/time.Second), 10))
			http.Error(w, "rate limit exceeded", http.StatusTooManyRequests)
			return
		}

		next.ServeHTTP(w, r)
	})
}

func rateKey(cfg *config.Config, r *http.Request) string {
	if tok := bearerToken(r.Header.Get("Authorization")); tok != "" && cfg.AdminKeyOK(tok) {
		return "admin:" + tok
	}
	if ip := clientIP(r); ip != "" {
		return "ip:" + ip
	}
	return "ip:unknown"
}

func clientIP(r *http.Request) string {
	if xff := r.Header.Get("X-Forwarded-For"); xff != "" {
		if i := strings.IndexByte(xff, ','); i >= 0 {
			return strings.TrimSpace(xff[:i])
		}
		return strings.TrimSpace(xff)
	}
	host, _, err := net.SplitHostPort(r.RemoteAddr)
	if err != nil {
		return r.RemoteAddr
	}
	return host
}

func bearerToken(h string) string {
	const p = "Bearer "
	if len(h) > len(p) && h[:len(p)] == p {
		return h[len(p):]
	}
	return ""
}

File: python_raalisence/middleware/ratelimit.py
"""Rate limiting middleware."""

import time
import threading
from typing import Dict, Optional
from fastapi import Request
from starlette.middleware.base import BaseHTTPMiddleware


class TokenBucket:
    """Token bucket rate limiter."""
    
    def __init__(self, rps: float, burst: int, ttl: int = 600):
        self.rps = rps  # tokens per second
        self.burst = burst  # max tokens
        self.ttl = ttl  # idle bucket eviction time
        self.tokens = float(burst)
        self.last_refill = time.time()
    
    def allow(self) -> tuple[bool, int, float]:
        """Check if request is allowed. Returns (allowed, remaining_tokens, retry_after)."""
        now = time.time()
        
        # Refill tokens
        elapsed = now - self.last_refill
        self.tokens = min(self.burst, self.tokens + elapsed * self.rps)
        self.last_refill = now
        
        if self.tokens >= 1.0:
            self.tokens -= 1.0
            return True, int(self.tokens), 0.0
        
        # Not enough tokens
        missing = 1.0 - self.tokens
        retry_after = missing / self.rps
        return False, int(self.tokens), retry_after
    
    def is_stale(self, now: float) -> bool:
        """Check if bucket is stale and should be evicted."""
        return now - self.last_refill > self.ttl


class RateLimiter:
    """Rate limiter with token bucket per client."""
    
    def __init__(self, rps: float, burst: int, ttl: int = 600):
        self.rps = rps
        self.burst = burst
        self.ttl = ttl
        self._lock = threading.Lock()
        self._buckets: Dict[str, TokenBucket] = {}
        self._last_sweep = time.time()
    
    def allow(self, key: str) -> tuple[bool, int, float]:
        """Check if request is allowed for given key."""
        with self._lock:
            now = time.time()
            
            # Periodic sweep of stale buckets
            if now - self._last_sweep > self.ttl:
                self._buckets = {
                    k: v for k, v in self._buckets.items() 
                    if not v.is_stale(now)
                }
                self._last_sweep = now
            
            # Get or create bucket
            bucket = self._buckets.get(key)
            if bucket is None:
                bucket = TokenBucket(self.rps, self.burst, self.ttl)
                self._buckets[key] = bucket
            
            return bucket.allow()


def rate_limit_key(request: Request, config) -> str:
    """Get rate limiting key for request."""
    # Check if request has valid admin token
    auth_header = request.headers.get("Authorization", "")
    if auth_header.startswith("Bearer "):
        token = auth_header[7:]
        if config.admin_key_ok(token):
            return f"admin:{token}"
    
    # Use client IP
    client_ip = request.client.host if request.client else "unknown"
    xff = request.headers.get("X-Forwarded-For")
    if xff:
        client_ip = xff.split(',')[0].strip()
    
    return f"ip:{client_ip}"


class RateLimitMiddleware(BaseHTTPMiddleware):
    """Rate limiting middleware."""
    
    def __init__(self, app):
        super().__init__(app)
        # Different limiters for different endpoints
        self.fast_limiter = RateLimiter(5.0, 10)  # validate/heartbeat
        self.admin_limiter = RateLimiter(1.0, 3)  # issue/revoke
        self.default_limiter = RateLimiter(2.0, 5)  # everything else
    
    def get_limiter(self, path: str) -> RateLimiter:
        """Get appropriate limiter for path."""
        if path in ["/api/v1/licenses/validate", "/api/v1/licenses/heartbeat"]:
            return self.fast_limiter
        elif path in ["/api/v1/licenses/issue", "/api/v1/licenses/revoke", "/api/v1/licenses/update", "/api/v1/licenses"]:
            return self.admin_limiter
        else:
            return self.default_limiter
    
    async def dispatch(self, request: Request, call_next):
        """Apply rate limiting."""
        # Get config from global variable
        try:
            from python_raalisence.server import config
            if config is None:
                # No config available, skip rate limiting
                response = await call_next(request)
                return response
        except ImportError:
            # Config not loaded yet, skip rate limiting
            response = await call_next(request)
            return response
        
        key = rate_limit_key(request, config)
        limiter = self.get_limiter(request.url.path)
        
        allowed, remaining, retry_after = limiter.allow(key)
        
        if not allowed:
            from fastapi.responses import Response
            return Response(
                content="rate limit exceeded",
                status_code=429,
                headers={
                    "Retry-After": str(int(retry_after)),
                    "RateLimit-Limit": "1",
                    "RateLimit-Remaining": str(remaining)
                }
            )
        
        response = await call_next(request)
        response.headers["RateLimit-Limit"] = "1"
        response.headers["RateLimit-Remaining"] = str(remaining)
        
        return response

